Update generic incompatibility notice

tarunvijwani opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #42072.

In #42724 PR, we've moved the switch to classic shortcode block button to separate components 

In this PR, we're updating the generic incompatibility notice message and adding the switch to classic shortcode block button.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Deactiavte any incompatible plugin like PayTm
2. Edit the Cart block page.
3. Focus on the Cart block. See the generic incompatibility notice in the sidebar. Click the "switch to classic cart" button.
4. Modal should open. Click switch. Block should be replaced.
5. In the snackbar bottom left, click "undo". Should revert back to blocks.
6. Repeat the steps for the Checkout Block page editor.

| Before | After |
|---|---|
|<img width="100%" alt="image" src="https://github.com/woocommerce/woocommerce/assets/11503784/c332bd1c-dc71-466f-b756-52e1e08ec0f0">|<img width="100%" alt="image" src="https://github.com/woocommerce/woocommerce/assets/11503784/7a21fa40-e048-4b8e-8488-d24523bd8e04">|

### For developer testing only 
1. Deactiavte any incompatible plugin like PayTm
2. Edit the Cart block page.
3. In browser console run `localStorage.debug = 'wc-admin:tracks';`
4. This will enable the logs for events in browser console. 
5. Focus on the Cart block. See the generic incompatibility notice in the sidebar. Click the "switch to classic cart" button.
4. Modal should open. Click switch. Block should be replaced.
5. Click on the undo button. 
6. In the browser console, confirm the following events are printed with `type: generic_notice`:
- wcadmin_switch_to_classic_shortcode_click
- wcadmin_switch_to_classic_shortcode_confirm
- wcadmin_switch_to_classic_shortcode_undo

![image](https://github.com/woocommerce/woocommerce/assets/11503784/637aaf5a-267e-41b3-a5cd-c4beac93caf8)


7. Enable PayTm or any incompatible plugin, confirm the notice is changed.
8. Focus on the Cart block. See the extension incompatibility notice in the sidebar. Click the "switch to classic cart" button.
9. Modal should open. Click switch. Block should be replaced.
10. Click on the undo button. 
11. In the browser console, confirm the following events are printed with `type: incompatible_notice`:
- wcadmin_switch_to_classic_shortcode_click
- wcadmin_switch_to_classic_shortcode_confirm
- wcadmin_switch_to_classic_shortcode_undo
<img width="1586" alt="image" src="https://github.com/woocommerce/woocommerce/assets/11503784/11eeae92-e5c2-447d-a04e-30cfae514428">

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Update the generic incompatibility notice message for the Cart and Checkout page.
 
#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Commit:
Move the switch to classic shortcode block button to separate component

- Move modal component file from incompatible-extension-notice component folder to switch-to-classic-shortcode-button folder.
- Rename modal component file to model-content component file.
- Move the switch to classic shortcode block button to separate component folder

Commit:
Update generic incompatibility notice

- Update the notice description.
- Add switch to classic shortcode block button to the notice.

Commit:
Fix margin in editor.scss

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

Commit:
Merge branch 'trunk' into update/42072-move-switch-classic-cart-checkout-button

Commit:
Merge branch 'update/42072-move-switch-classic-cart-checkout-button' of https://github.com/woocommerce/woocommerce into update/42072-move-switch-classic-cart-checkout-button

Commit:
Update class names in switch-to-classic-shortcode-button component

Commit:
Merge branch 'update/42072-move-switch-classic-cart-checkout-button' into update/42072-update-incompatibility-generic-notice

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

Commit:
Merge branch 'trunk' into update/42072-move-switch-classic-cart-checkout-button

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

Commit:
Update plugins/woocommerce-blocks/assets/js/editor-components/switch-to-classic-shortcode-button/editor.scss

Co-authored-by: Niels Lange <info@nielslange.de>

Commit:
Refactor SwitchToClassicShortcodeButton component

Commit:
Fix css lint erros

tarhi-saad reviewed (changes requested):
Thank you, @tarunvijwani, for working on this PR! The testing works as expected! I only have a couple of questions:

- The text suggested by Pedro in #42072 is: `Some extensions don't yet support this block, which may impact the shopper experience. To make sure this feature is right for your store, [review the list of compatible extensions](https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-3).`. However, I noticed that the wording and the link in this PR are not the same. Was there a decision made to change the wording?
- I have a different design in my testing for the incompatibility notice. Do you know what could be the reason?
<img width="288" alt="image" src="https://github.com/woocommerce/woocommerce/assets/14235870/9a6a9f8a-ac40-4f50-ad5d-bef579608c06">

- There was another task mentioned in the ticket related to this PR:

> Add a new property to the following events to distinguish when the flow was triggered by the generic notification or the one that outlines the incompatible extensions: wcadmin_switch_to_classic_shortcode_click, wcadmin_switch_to_classic_shortcode_confirm, wcadmin_switch_to_classic_shortcode_undo, wcadmin_switch_to_classic_shortcode_cancel.

Commit:
Merge branch 'update/42072-move-switch-classic-cart-checkout-button' into update/42072-update-incompatibility-generic-notice

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

Commit:
Add type property to notice events

tarunvijwani commented:
> Thank you, @tarunvijwani, for working on this PR! The testing works as expected! I only have a couple of questions:
> 
> * The text suggested by Pedro in https://github.com/woocommerce/woocommerce/issues/42072 is: `Some extensions don't yet support this block, which may impact the shopper experience. To make sure this feature is right for your store, [review the list of compatible extensions](https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-3).`. However, I noticed that the wording and the link in this PR are not the same. Was there a decision made to change the wording?
> * I have a different design in my testing for the incompatibility notice. Do you know what could be the reason?
>
> <img alt="image" width="288" src="https://private-user-images.githubusercontent.com/14235870/291180132-9a6a9f8a-ac40-4f50-ad5d-bef579608c06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI5MTMxNTAsIm5iZiI6MTcwMjkxMjg1MCwicGF0aCI6Ii8xNDIzNTg3MC8yOTExODAxMzItOWE2YTlmOGEtYWM0MC00ZjUwLWFkNWQtYmVmNTc5NjA4YzA2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE4VDE1MjA1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyNmMxNTc0NWM1NGU1YjYyMzNjOGZjNzFkNWUwMGFiZWQ2ODhmYzhiNWJmYTNkMDRkMmYwYmE5NzAzMTY5YTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fAzUNIr4itzoEuaCZizeCldV2CQWiPn_3eCBQSFONaE">

@tarhi-saad Thank you for testing the PR. It seems you have an incompatible plugin active on your site. You will need to deactivate all incompatible plugins to see the generic notice. 
 
> * There was another task mentioned in the ticket related to this PR:
> 
> > Add a new property to the following events to distinguish when the flow was triggered by the generic notification or the one that outlines the incompatible extensions: wcadmin_switch_to_classic_shortcode_click, wcadmin_switch_to_classic_shortcode_confirm, wcadmin_switch_to_classic_shortcode_undo, wcadmin_switch_to_classic_shortcode_cancel.



Thank you, I almost forgot this part. I have updated the code and added the developer testing steps for it in the description.
Please take a moment to review the PR again and let me know if any other changes are required.

Commit:
Merge branch 'trunk' into update/42072-update-incompatibility-generic-notice

tarunvijwani commented:
Moving the PR to draft for code improvements and optimization.

Commit:
Refactor switch to classic shortcode button logic

Commit:
Use isCart for switchButtonLabel and snackbarLabel

Commit:
Remove empty div in sidebar compatibility notice component

Commit:
Refactor switch to classic shortcode button event handling

wavvves commented on plugins/woocommerce-blocks/assets/js/editor-components/sidebar-compatibility-notice/index.tsx:
```diff
@@ -11,17 +11,26 @@ import classnames from 'classnames';
 	const noticeText = createInterpolateElement(
 		__(
-			'The Cart & Checkout Blocks are built to optimize for faster checkout. To make sure this feature is right for your store, <a>review the list of compatible extensions</a>.',
+			"Some active extensions don't yet support this block, which may impact the shopper experience. <a>Learn more</a>",
```
Are we sure we want to include the word "active" when no incompatible extensions are activated? I might be wrong but, by reading the issue, it seems to me this was a tiny mistake turned into truth along the way 🤔

@pmcpinto can you have a glance real quick on this?

nielslange replied:
> Are we sure we want to include the word "active" when no incompatible extensions are activated? I might be wrong but, by reading the issue, it seems to me this was a tiny mistake turned into truth along the way 🤔

@wavvves We're only checking if activated extensions are incompatible. That's both due to technical- and performance-reasons. Below, you can find screenshots that show that the incompatible notice only appears when at least one incompatible extension has been activated. 

### Incompatible extension(s) installed, but not activated

<img width="1800" alt="Screenshot 2023-12-28 at 10 51 01" src="https://github.com/woocommerce/woocommerce/assets/3323310/0d4051d7-d19b-4008-a988-42f0ee9e805b">

### Incompatible extension(s) installed, and at least one activated

<img width="1800" alt="Screenshot 2023-12-28 at 10 51 24" src="https://github.com/woocommerce/woocommerce/assets/3323310/aa968b3c-964d-40aa-9c31-4c5883dae586">

wavvves replied:
Correct me if I'm wrong, but this PR isn't about checking for activated incompatible extensions. It's for changing the generic message and including an easy path for shortcode reversion. At least this is what I get from reading the issue https://github.com/woocommerce/woocommerce/issues/42072

Now, in the issue, the plan is to have the message:
> Some extensions don't yet support this block, which may impact the shopper experience. To make sure this feature is right for your store, [review the list of compatible extensions](https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-3).

This is the generic message, just stating there are extensions out there not working with Blocks, and should be shown without incompatible extensions activated.

In this https://github.com/woocommerce/woocommerce/issues/42072#issuecomment-1850735158 the word "active" is introduced, and that seems to have been replicated here. 

Also, the phrase was changed as noted by @tarhi-saad here https://github.com/woocommerce/woocommerce/pull/42751#pullrequestreview-1785986966:

> The text suggested by Pedro in https://github.com/woocommerce/woocommerce/issues/42072 is: Some extensions don't yet support this block, which may impact the shopper experience. To make sure this feature is right for your store, [review the list of compatible extensions](https://woocommerce.com/document/cart-checkout-blocks-support-status/#section-3).. However, I noticed that the wording and the link in this PR are not the same. Was there a decision made to change the wording?

nielslange replied:
Good catch, @wavvves, and apologies my earlier comment. The word `active` should indeed not be part of the generic message. If there is at least one incompatible extension activated, another message will be displayed. I've adjusted this accordingly.

wavvves committed:
Merge branch 'trunk' into update/42072-update-incompatibility-generic-notice

nielslange committed:
Adjust wording

nielslange committed:
Adjust marging and readability

wavvves commented on plugins/woocommerce/changelog/42751-update-42072-update-incompatibility-generic-notice:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: enhancement
+
+Update the generic incompatibility notice message for the Cart and Checkout page.
```
do we need a changelog entry here too? 🤔

nielslange replied:
@wavvves Sorry for answering this question delayed. The changelog entry in `/plugins/woocommerce/` is required.

wavvves commented on plugins/woocommerce-blocks/assets/js/editor-components/sidebar-compatibility-notice/index.tsx:
```diff
@@ -33,6 +42,9 @@ export const CartCheckoutSidebarCompatibilityNotice = ( {
 	);
 
+	const blockName =
+		block === 'cart' ? 'woocommerce/cart' : 'woocommerce/checkout';
```
wouldn't it be safer to concatenate "woocommerce/" with the block name an let it error out if something invalid appears here?

The way it is now, if a invalid block slips into here, it will assume woocommerce/checkout and be silent about the invalid data.

nielslange replied:
`block` is a props of the `<CartCheckoutSidebarCompatibilityNotice>` component. The only place, we're using this component is on https://github.com/woocommerce/woocommerce/blob/2749cebf7f06b5dbfd45550a8dc4f65038ff9bb8/plugins/woocommerce-blocks/assets/js/blocks/cart-checkout-shared/sidebar-notices/index.tsx#L133-L135

Technically, it should not be possible that `block` has a value apart from `cart` or `checkout`. Also, the following part, which is located before the part I mentioned before, should prevent that anything other than the Cart or the Checkout block every reaches this section: https://github.com/woocommerce/woocommerce/blob/2749cebf7f06b5dbfd45550a8dc4f65038ff9bb8/plugins/woocommerce-blocks/assets/js/blocks/cart-checkout-shared/sidebar-notices/index.tsx#L109-L115

That said, may I ask what scenario you have in mind? How do you think an invalid block might slip into here?

wavvves replied:
I'm not saying it can have a value not in [cart, checkout] right now, but in the future a new value is introduced, the ternary will erroneously default to woocommerce/checkout with probably no immediate and visible side effects.

nielslange replied:
I see what you mean, @wavvves. I believe that if we introduce a new value in the future, the TS types should be adjusted accordingly. The part, that you pointed out is located at

https://github.com/woocommerce/woocommerce/blob/27899be1549d2d3670ca1e7527a5a88fbbc499c8/plugins/woocommerce-blocks/assets/js/editor-components/sidebar-compatibility-notice/index.tsx#L45-L46

It's part of the `CartCheckoutSidebarCompatibilityNotice`, which is using this TS interface

https://github.com/woocommerce/woocommerce/blob/27899be1549d2d3670ca1e7527a5a88fbbc499c8/plugins/woocommerce-blocks/assets/js/editor-components/sidebar-compatibility-notice/index.tsx#L16-L19

Initially, you mentioned 
> wouldn't it be safer to concatenate "woocommerce/" with the block name an let it error out if something invalid appears here?

What would be your proposal in this case? Something like this:

```ts
const blockName = `woocommerce/${ block }`;
if (
	blockName !== 'woocommerce/cart' &&
	blockName !== 'woocommerce/checkout'
) {
	throw new Error( 'Invalid blockName' );
}
```

wavvves replied:
If we're only accepting 'cart' | 'checkout' earlier, then `const blockName = 'woocommerce/${ block }';` will suffice right? Any other value specified earlier will be rejected at interface level, and that if is unneeded and unreachable.

If the interface changes to be less strict, then it would fail later after concatenation, because it wouldn't be able to find blockName.

I don't want to delay this PR, but IMO that would be better than a ternary that defaults the value. Let it fail, imagine in the future you are doing some task, introduce an invalid block at interface level by distraction (typo etc), it defaults to checkout and you never realise it because it will never complain about it.

nielslange replied:
Thanks for clarifying this, @wavvves. I've updated the part you mentioned and went a step ahead. Instead of using
```ts
const blockName = `woocommerce/${ block }`;
```
and
```ts
<SwitchToClassicShortcodeButton
    block={ blockName }
    clientId={ clientId }
    type={ 'generic' }
/>
```
I'm directly calling
```ts
<SwitchToClassicShortcodeButton
    block={ `woocommerce/${ block }` }
    clientId={ clientId }
    type={ 'generic' }
/>
```

Please let me know what you think about that approach.

wavvves commented on plugins/woocommerce-blocks/assets/js/editor-components/switch-to-classic-shortcode-button/index.tsx:
```diff
@@ -35,15 +37,25 @@ export function SwitchToClassicShortcodeButton( {
+		: __( 'Switched to classic checkout.', 'woocommerce' );
+
+	const notice =
+		type === 'incompatible' ? 'incompatible_notice' : 'generic_notice';
```
Same here, invalid type data won't produce an error and might display the generic notice erroneously. This way the bug could be in the wild for a long time before it gets reported.

nielslange replied:
`type` is a props of the `<SwitchToClassicShortcodeButton>` component. Currently, we're only using this component here: https://github.com/woocommerce/woocommerce/blob/2749cebf7f06b5dbfd45550a8dc4f65038ff9bb8/plugins/woocommerce-blocks/assets/js/editor-components/incompatible-extension-notice/index.tsx#L144-L147

Assuming that we'd merge this PR, we'd end up with these two usages in our codebase:

```ts
<SwitchToClassicShortcodeButton
  block={ block }
  clientId={ clientId }
  type={ 'incompatible' }
/>
```

```ts
<SwitchToClassicShortcodeButton
  block={ blockName }
  clientId={ clientId }
  type={ 'generic' }
/>
```

That said, we could safe guard the `type` prop by changing

```ts
interface SwitchToClassicShortcodeButtonProps {
	block: 'woocommerce/cart' | 'woocommerce/checkout';
	clientId: string;
	type: string;
}
```

to

```ts
interface SwitchToClassicShortcodeButtonProps {
	block: 'woocommerce/cart' | 'woocommerce/checkout';
	clientId: string;
	type: 'incompatible' | 'generic';
}
```

That way, we'll imediately see an TS error is providing an invalid type. I'll adjust this accordingly.

nielslange committed:
Introduce union for SwitchToClassicShortcodeButtonProps type prop

nielslange committed:
Simplify logic

wavvves commented on plugins/woocommerce-blocks/assets/js/editor-components/sidebar-compatibility-notice/index.tsx:
```diff
@@ -11,17 +11,26 @@ import classnames from 'classnames';
 	const noticeText = createInterpolateElement(
 		__(
-			'The Cart & Checkout Blocks are built to optimize for faster checkout. To make sure this feature is right for your store, <a>review the list of compatible extensions</a>.',
+			"Some extensions don't yet support this block, which may impact the shopper experience. <a>Learn more</a>",
```
I'm sorry Niels 😅, are we supposed to drop the "review the list yada yada" part? What we got now mismatches the issue acceptance criteria https://github.com/woocommerce/woocommerce/issues/42072

<img width="319" alt="image" src="https://github.com/woocommerce/woocommerce/assets/17236129/b01ac42f-74cb-4e0a-9a39-1d7eca6bd35d">

nielslange replied:
Thanks for catching that, @wavvves. You're correct. According to the acceptance criteria in #42072, the current message was incorrect. I just corrected that. You can now review this PR again.

wavvves committed:
Merge branch 'trunk' into update/42072-update-incompatibility-generic-notice

nielslange committed:
Use proposed text

nielslange commented:
Thanks for reviewing and approving this PR, @wavvves. 🙌
